engine/crates/engine-v2/src/plan/planning/boundary.rs

```diff
@@ -0,0 +1,526 @@
+        self.planner
+    }
+}
+
+impl<'schema, 'a> std::ops::DerefMut for BoundaryPlanner<'schema, 'a> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.planner
+    }
+}
+
+/// During the planning of the boundary we need to keep track of fields by their FieldId
+/// to satisfy requirements. The goal is not only to know what's present but also to have the
+/// correct ResponseEdge for those when reading data from the response later.
+type BoundaryFields = FnvHashMap<FieldId, BoundaryField>;
+
+#[derive(Debug)]
+struct BoundaryField {
+    plan_id: PlanId,
+    /// BoundaryField with the lowest position (lowest edge) within the query
+    /// for a give FieldId
+    final_bound_field_id: BoundFieldId,
+    subselection_set_ids: Vec<BoundSelectionSetId>,
+    lazy_subselection: Option<BoundaryFields>,
+}
+
+impl BoundaryField {
+    fn new(plan_id: PlanId, group: GroupedByFieldId) -> Self {
+        Self {
+            plan_id,
+            final_bound_field_id: group.final_bound_field_id,
+            subselection_set_ids: group.subselection_set_ids,
+            lazy_subselection: None,
+        }
+    }
+}
+
+/// Potential child plan, but might not be the best one.
+struct ChildPlanCandidate<'schema> {
+    resolver_id: ResolverId,
+    /// Entity type (object/interface id) of the fields
+    entity_type: EntityType,
+    /// Providable fields by the resolvers with their requirements
+    providable_fields: Vec<(BoundFieldId, &'schema FieldSet)>,
+}
+
+/// Field that the parent plan could not providabe.
+struct UnplannedField<'schema> {
+    entity_type: EntityType,
+    flat_field: FlatField,
+    field_resolvers: Vec<FieldResolverWalker<'schema>>,
+}
+
+impl<'schema> std::ops::Deref for UnplannedField<'schema> {
+    type Target = FlatField;
+    fn deref(&self) -> &Self::Target {
+        &self.flat_field
+    }
+}
+
+impl<'schema> From<UnplannedField<'schema>> for FlatField {
+    fn from(unplanned: UnplannedField<'schema>) -> Self {
+        unplanned.flat_field
+    }
+}
+
+impl<'schema, 'a> BoundaryPlanner<'schema, 'a> {
+    /// Iteratively plan fields.
+    /// 1. Generate all potential plan candidates satisfying their requirements if possible.
+    /// 2. Select the best candidate, generate its input and remove its output fields from the
+    ///    unplanned ones.
+    /// 3. Continue until there are no more unplanned fields.
+    fn plan_selection_set(
+        mut self,
+        boundary_fields: &mut BoundaryFields,
+        mut unplanned_selection_set: FlatSelectionSet,
+    ) -> PlanningResult<Vec<PlanId>> {
+        // Fields that couldn't be provided by the parent and that have yet to be planned by one
+        // child plan.
+        let mut id_to_unplanned_fields: FnvHashMap<BoundFieldId, UnplannedField<'schema>> = self
+            .build_unplanned_fields(std::mem::replace(
+                &mut unplanned_selection_set.fields,
+                Vec::with_capacity(0),
+            ));
```

Comment:
Is it not easier to take here?

```suggestion
            .build_unplanned_fields(std::mem::take(
                &mut unplanned_selection_set.fields,
            ));
```

Reply:
yeah definitely. Thought I was smart using `Vec::with_capacity(0)` when `Vec::new()` doesn't allocate until the first push 😓